src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -782,13 +782,8 @@ public LayoutInt CalculateFieldBaseOffset(MetadataType type, bool requiresAlign8
             if (!type.IsValueType && type.HasBaseType)
             {
                 cumulativeInstanceFieldPos = type.BaseType.InstanceByteCountUnaligned;
-                if (!type.BaseType.InstanceByteCountUnaligned.IsIndeterminate)
+                if (!cumulativeInstanceFieldPos.IsIndeterminate)
                 {
-                    cumulativeInstanceFieldPos = type.BaseType.InstanceByteCountUnaligned;
-                    if (type.BaseType.IsZeroSizedReferenceType && ((MetadataType)type.BaseType).HasLayout())
```

Comment:
Can `IsZeroSizedReferenceType` propety be deleted as well? It does not appear to be used anywhere after this change.

Reply:
I suspect we still need to do something about the zero-sized layout types because runtime artificially bumps up their size to 1 even though it rolls back the change when calculating the layout for a derived type. I guess that, as a first step, I need to see how the Pri1 tests look on the different architectures after my change. Other than that, I was under the impression that <code>IsZeroSizedReferenceType</code> was a pre-existing NativeAOT construct, not a Crossgen2-specific addition, but I may be mistaken about this.

Reply:
IsZeroSizedReferenceType was added in https://github.com/dotnet/runtime/pull/46769

Reply:
Thanks for sharing the link to the original PR. ARM tests are already failing in the Pri1 tests exactly because of the missing zero-sized layout compensation, I believe the trick is not about removing the check but figuring out the proper place for it, that's my next step.